### PR TITLE
Use RELATIVE_PROJECT_PATH in export

### LINKED ISSUE #25: Workflow action doesn't read export path
We have a game were we want to automatically deploy it with github pages, but we're encountering an issue regarding export path.
```
rsync: change_dir "/home/runner/work/crackout/crackout//HTML5/" failed: No such file or directory (2)
````
is the error we get, because this is the export path we set in export_presets.cfg.
But this is not the path were our game is exported. If we go up we can see it's exported here
```
/home/runner/.local/share/godot/godot_executable/Godot_v3.2.2-stable_mono_linux_headless_64/godot /home/runner/work/crackout/crackout/project.godot --export HTML5 /home/runner/.local/share/godot/builds/HTML5/index.html --verbose
```
More specifically here: `/home/runner/.local/share/godot/builds/HTML5/index.html`
While we expect it to be exported to our chosen path here:
`/home/runner/work/crackout/crackout/HTML5/` (Rather have it at `home/runner/crackout/HTML5` but that's another issue)

Here is our deployment.yml
```
name: Deployment

on:
  push:
    branches-ignore:
      - gh-pages

jobs:
  export_game:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout
      uses: actions/checkout@v2
    - name: Export
      uses: firebelley/godot-export@v2.2.0
      with:
        godot_executable_download_url: https://downloads.tuxfamily.org/godotengine/3.2.2/mono/Godot_v3.2.2-stable_mono_linux_headless_64.zip
        godot_export_templates_download_url: https://downloads.tuxfamily.org/godotengine/3.2.2/mono/Godot_v3.2.2-stable_mono_export_templates.tpz
        relative_project_path: ./
        create_release: false
    - name: Deploy
      uses: JamesIves/github-pages-deploy-action@releases/v3
      with:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        BRANCH: gh-pages
        FOLDER: ../../../.local/share/godot/builds/HTML5/
        TARGET_FOLDER: ${GITHUB_REF##*/}/game/
        CLEAN: true
```
Now, in this yml we include the fix which is rather disgusting `../../../.local/share/godot/builds/HTML5/`. And well, frankly this is not desired. Our export_presets.cfg looks like this: 
```
[preset.0]

name="HTML5"
platform="HTML5"
runnable=true
custom_features=""
export_filter="all_resources"
include_filter=""
exclude_filter=""
export_path="HTML5/index.html"
patch_list=PoolStringArray(  )
script_export_mode=1
script_encryption_key=""

[preset.0.options]

vram_texture_compression/for_desktop=true
vram_texture_compression/for_mobile=false
html/custom_html_shell=""
html/head_include=""
custom_template/release=""
custom_template/debug=""
```

### PR DESCRIPTION
**Pull request by BluePhoenixGame**
Here we assume preset.export_path is relative to the project path, which should be the case since this is set up in the editor were the project is the root.
Fixes #25, no longer needing `../../../.local/share/godot/builds/` to reach it during deployment.

### CODE REVIEW

File: src/godot.ts [hunk 1]
```diff
@@ -151,7 +151,7 @@ async function doExport(): Promise<BuildResult[]> {
 
     let executablePath;
     if (preset.export_path) {
-      executablePath = path.join(buildDir, path.basename(preset.export_path));
+      executablePath = path.join(RELATIVE_PROJECT_PATH, preset.export_path);
```
Comment (firebelley) [inline src/godot.ts:154]: I'm afraid this isn't going to work as cleanly as you might hope. Let's say a user has their export path defined simply as `windows.exe`. What this will do, is build the `windows.exe` and `game.pck` at the same level as `project.godot`. Then, when a release is created it will zip up ALL of the contents of the folder, including all of the raw development resources. This was the reason for building the games to a separate builds directory, so that the entirety of the directory could be safely zipped in its entirety.

You could approach this in a couple of ways:
- You can check to make sure the directories the builds are going to are empty (you'd also have to make sure that multiple presets aren't all pointing to the same directory)
- You can keep the builds going to a separate directory and modify [this function](https://github.com/firebelley/godot-export/blob/71657a8f8e3b4021f531ae9362d9966dfb47d52c/src/file.ts#L37) to behave slightly differently

Alternatively we could just deal with the potential for user error, but I fear that this functionality won't be totally clear and will result in some undesirable effects. Also, I'd like to see some way to control this behavior via an optional input to the action.

Looking forward to hearing back from you! Thanks again for the contribution.

Comment (BluePhoenixGame) [reply src/godot.ts:154]: Ah yes, I see my issue. I think I have an idea though, we could by default set `relative_export_path` to be the export_path. So in my case where I have `export_path="HTML5/index.html"`, if I leave `relative_export_path` empty it would be equal to `HTML5/`.
I wanna know what you think of this idea.

Comment (firebelley) [reply src/godot.ts]: Yes this sounds like a good approach, I think the default behavior should still be to keep builds in that transient `builds` directory unless the user specifies that they want the builds exported in accordance with the paths in `export_presets.cfg`. Keeping the builds in that `builds` directory just makes a lot of operations easier.

Maybe a new optional input, something like `move_exports` or `relative_exports` that when set to `true` will use the `export_presets.cfg` paths relative to the project root. That makes sense to me. I'm not super concerned with the details, I just think it would be nice to control that behavior. I personally don't have a use-case currently where I want my exports to exist anywhere in particular, I just want builds for my release.

Something else to consider is there is an input where you can archive all the outputs: `archive_export_output`, we'll want to make sure this is adhered to still when using the path present in `export_presets.cfg`.

Just some context for you to consider, happy to talk about this further but I think overall your suggestion is a good one!

**Comment by BluePhoenixGame** [comment]
Okay, I think I found the best way to implement this. I am also happy with the documentation.
But you needed me to build with yarn? Let me clone this and do that.

**Comment by firebelley** [comment]
Awesome! Yeah if you aren't familiar with Yarn you can check it out here `https://yarnpkg.com/`

Once you have it installed do a `yarn install` then `yarn build` and commit the output file. Then once that's good to go, I can do some testing and review, but so far the draft looks great!

**Comment by BluePhoenixGame** [comment]
That should be all!
_I had to fix two linting problems ESLint didn't like (mostly regarding line length)._

Should be all done now. But there is one thing I would like to discuss. And that is changing this new constant to default to true. I set it to false now while working but I think it should actually be changed.
**Pros**: Expected behaviour since you define it in `export_presets.cfg`. You don't repeat yourself because of this, good design.
**Cons**: Breaking change because this new constant will overwrite and be prioritized over `relative_export_path`. This could be changed thus making `relative_export_path` higher in priority.

**Comment by BluePhoenixGame** [comment]
Thinking through the pros and cons I realized that it could be good to document the current bias.

**Comment by firebelley** [comment]
So I have a few thoughts, totally open to doing what you want to do but here are some things for consideration:
- Defaulting the new property to `true` is a breaking change, which would require a major version upgrade (not a big deal but something to consider)
  - Maybe we should consider getting rid of the `relative_export_path` altogether if we want to make this a major version upgrade, and just inform users that they'll have to use the path in export presets
  - This could be an issue for users with lots of presets (I have 8 configured presets on one of my projects!)
  - Perhaps the `relative_export_path` could be repurposed as yet another location where all builds get moved? So users could have the option of dumping all builds in a single location regardless of what's configured in their export presets?
- If we want to keep this to a minor version upgrade, I think it pretty much has to be the case that the default value is of the new property is `false` so that existing users can upgrade without "breaking" changes
  - In this case I think it's definitely smart to document that the new property takes precedence over relative_export_path
  - Keeping this to a minor version upgrade doesn't preclude us from making this default behavior in the future with a major version upgrade. This might be preferable because at the moment I am kind of in the dark about how everyone is using this action and so can't make informed decisions about the best default user experience.

Basically, if we want to get this change in without too many changes we'd have to keep the value defaulted to false and do a minor version upgrade. However, if you'd be OK with me (or you!) taking the opportunity to rework some things as part of a major version upgrade I would be totally open to that, with the caveat that this feature would then go out at some point in the future. And then there's a middle road where this goes in now, and becomes default behavior in a future version.

I'm totally fine with whatever decision you choose to make, so just let me know! I'll be leaving a comment or two on the code and I'll have time to test the change tomorrow.

Nice work, and thanks again for the contribution!

File: src/constants.ts [hunk 2]
```diff
@@ -12,6 +12,7 @@ const RELATIVE_EXPORT_PATH = core.getInput('relative_export_path');
 const RELATIVE_PROJECT_PATH = core.getInput('relative_project_path');
 const SHOULD_CREATE_RELEASE = core.getInput('create_release') === 'true';
 const UPDATE_WINDOWS_ICONS = core.getInput('update_windows_icons') === 'true';
+const USE_PRESET_EXPORT_PATH = core.getInput('use_preset_export_path') === 'false';
```
Comment (firebelley) [inline src/constants.ts:15]: This does not set a default value, rather this just converts the inputs to boolean values for use later. This line should be `=== 'true'` like the others, and the input and its default value need to be defined in [this file](https://github.com/firebelley/godot-export/blob/master/action.yml). Let me know if anything is unclear!

File: src/file.ts [hunk 3]
```diff
@@ -35,13 +35,15 @@ async function zipBuildResult(buildResult: BuildResult): Promise<void> {
...
 async function moveBuildsToExportDirectory(buildResults: BuildResult[], moveArchived?: boolean): Promise<void> {
-  const fullExportPath = path.resolve(RELATIVE_EXPORT_PATH);
-  core.startGroup(`Moving exports to ${fullExportPath}`);
-
-  await io.mkdirP(fullExportPath);
-
   const promises: Promise<void>[] = [];
   for (const buildResult of buildResults) {
+    const fullExportPath = path.resolve(
+      USE_PRESET_EXPORT_PATH ? path.dirname(buildResult.preset.export_path) : RELATIVE_EXPORT_PATH,
```
Comment (firebelley) [inline src/file.ts:41]: Definitely will want to document that the new property takes precedence over `relative_export_path`

Comment (BluePhoenixGame) [inline src/file.ts:41]: @firebelley: "This could be an issue for users with lots of presets (I have 8 configured presets on one of my projects!)"
What would be the problem here? I moved the path inside the loop so each export gets moved to where it is defined. This is even an improvement because before you couldn't move them to different paths.

Comment (BluePhoenixGame) [reply src/file.ts:41]: Yes, I realized this too and fixed it in the most recent commit 3f0aed2025956d904bb1ef439d214f59de970716

Comment (firebelley) [reply src/file.ts:41]: I'm saying it would a problem in general if I had a large number of export presets and I wanted them to all go to the same place, I'd have to manually change the paths for each one of them. That's a non-issue if we're keeping `relative_export_path`

Comment (BluePhoenixGame) [reply src/file.ts:41]: Ah, okay. But I think that it's not really a problem at all. The action already requires you to set a path in the export_presets. So it shouldn't be much additional work if we think from a new users perspective. I do see the minor problem of migrating to this. But overall I don't think that this is big enough of a problem as it's not intense manual work but rather adding that folder to the path in export_presets.

**Comment by BluePhoenixGame** [comment]
Regarding breaking change, I say we merge it like this and I'll open another pull request that changes the default and removes `relative_export_path` that we can merge when a complete new version is about to be published (3.x)

**Comment by firebelley** [comment]
Ok, sounds good! Please make sure you've built your latest changes with `yarn build` and I'll do some testing tonight. If all goes well I'll go ahead and merge the change and create a new version 👍

**Comment by BluePhoenixGame** [comment]
Alright, I built again!

File: src/file.ts [hunk 4]
```diff
@@ -35,13 +35,15 @@ async function zipBuildResult(buildResult: BuildResult): Promise<void> {
...
+    );
+    core.startGroup(`Moving exports to ${fullExportPath}`);
```
Comment (firebelley) [inline src/file.ts:43]: Just this minor thing, can you move this `core.startGroup` to outside the loop? The message can be changed to something like `'Moving exports'` instead. Also friendly reminder to `yarn build` again once that's been changed.

**Comment by BluePhoenixGame** [comment]
Fixed that spammy message now, cd9b65299ddf98411b004684ccc7a48e8f025bc9 includes build.
